Encryption.md

```diff
@@ -19,216 +19,544 @@
+   and connect all readers to a KMS. The footer key is generated as F(key_metadata), where F is an 
+   algorithm unknown outside the organization. These keys are less secure than KMS-managed keys, but 
+   better than using no keys at all in plaintext and unsigned footers (discussed in the sections 5.2 
+   and 5.3.)
+   
+Key metadata can also be empty - in a case the encryption keys are fully managed by the caller 
+code, and passed explicitly to Parquet for the footer and each encrypted column.
+
+### 4.4 Additional Authenticated Data
+The AES GCM cipher in the basic mode protects against byte replacement inside a ciphertext, 
+but can't prevent replacement of one ciphertext with another (encrypted with the same key). 
+This can be solved by the AAD mode, using different AADs for different ciphertexts. Parquet 
+modular encryption uses the AAD mode to protect against swapping ciphertext modules inside 
+a file, between files - or against swapping full files (for example, replacement of a file 
+with an old version). Obviously, the AAD must reflect the identity of a file and of the modules 
+inside the file.
+
+Parquet constructs a module AAD from two components: an optional AAD_Prefix - a string provided 
+by the user for the file, and an AAD_Suffix, built internally for each GCM-encrypted module 
+inside the file. The AAD prefix should reflect the target identity that helps to detect file 
+swapping (simple example - table name with a date and partition, e.g., "employees_23_05_2018.part0"). 
+The AAD suffix reflects the internal identity of modules inside the file, which for example 
+prevents replacement of column pages in row group 0 by pages from the same column in row 
+group 1 (when columns are encrypted with the CGM cipher). The module AAD is a direct concatenation 
+of the prefix and suffix parts.
+
+#### 4.4.1 AAD_Prefix
+A file writer passes an optional AAD_Prefix string upon file creation, that allows to differentiate 
+it from other files. The reader should know this string apriori, or should be able to 
+retrieve and verify it, using a convention accepted in the organization. An optional `aad_metadata` 
+byte array can be specified in a file - to help a reader retrieve the AAD 
+prefix string used for file enciphering. The aad_metadata  field can be empty, since readers might 
+know apriori a right AAD_Prefix for each file. Alternatively, the aad_metadata  can contain hints 
+that help readers to get the right AAD prefix. Yet another option is to keep the AAD_Prefix itself 
+inside the aad_metadata field – in that case, the readers must be able to verify it independently, 
+since this field is not protected in a storage, and can be changed or filled by an attacker.
+
+Here go a number of examples of AAD_Prefix creation and management in an organization:
+
+   * An organization uses a convention is to build an AAD_Prefix from the table name/ID, date and 
```

Comment:
Quite a few examples here, you might consider removing some of them.

Reply:
Proper usage of AAD prefixes and aad_metadata might be challenging, so having a number of diverse examples is helpful.

Reply:
I think I can propose a major simplification of the AAD prefix management. Users won't have to do anything, except for make the writer generate and pass the AAD prefix (and make readers verify it belongs to a right dataset). Most examples become irrelevant, since Parquet will store and fully manage the AAD prefix itself (we'll still need an explanation on AAD prefix generation and verification).

Let me know if I miss anything:
AAD prefix can be stored inside the aad_metadata field. If an attacker replaces it, the file modules won't be readable, so its ok (files can always be corrupted, in many ways). However, if its stored as a plaintext, the attacker can learn the file identity (eg "customer table") and infer some info from its size (eg decreases from month to month) or date of modification. But we can encrypt the AAD prefix with the footer key (with CTR, or GCM without using an AAD), and store it in the `aad_metadata` field (to be renamed to `aad_prefix`). 

This is for encrypted footer mode. An open question is about plaintext footer mode. The file writers have a footer key (for footer signature creation), so can encrypt the AAD prefix. But this will require the readers of sensitive columns to have the footer signing key. Today its optional, because a reader might want to skip footer signature verification. We have two options:
- require readers of sensitive columns to always have the footer signing key (and always check footer integrity, which might be a good thing to do in any case)
- don't encrypt AAD prefix for files with plaintext footer. The footer fields are visible anyway, hence the attacker knows the schema etc anyway and can guess the file identity.

Reply:
Tomorrow, I'll start pushing the first pull requests in this round. They won't include anything we haven't agreed upon.  @rdblue @zivanfi let me know if the proposal above is ok with you. I believe the AADs are the topic that will be the hardest to grasp by future users. While we already handle the suffix part internally, it will be good to handle management of the prefix part too, so the users can focus only on its generation and verification, not on management of its storage or encryption.

Reply:
I think that encrypting the AAD prefix sounds over-complicated. It requires using GCM in a different way or using CTR when everything else is GCM. That adds a lot of complexity to the spec. This also creates the issues that you outlined for plaintext footer mode.

I would prefer a simpler approach. The reason you are trying to protect the AAD prefix is assuming that it gives away information and identifies the file. If that's the case, then the user should track it separately. If it doesn't give away identifying information, then it is okay to store in plain text.

I think the easiest way to implement this is to have a flag for whether the prefix is stored in the file or not. If it isn't, then we throw an exception if it isn't supplied. If it is, then the user can query it and validate it matches the expected prefix.

We can change this later to a more complicated scheme, but I think it is a reasonable start. I'd rather do something simple and get this spec finished than spend a lot of time designing an optional feature like this one right now. What do you think?

Reply:
Sounds good. Actually, the files/folders are often named after the table identity, so encrypting the AAD prefix won't be sufficient. So, to sum this up:
- aad_metadata field is renamed to aad_prefix. AAD prefix (if passed by writer) will be stored there.  This removes prefix management complexity from basic  users. 
- Parquet will have a writer API option not to store the aad_prefix - for advanced users with high secrecy requirements (or basic users with advanced  frameworks). These users/frameworks will   track  prefixes separately, and also handle generation of eg random file names, along with overall table metadata management.

Question: do we need an explicit flag on whether the prefix is stored - or can use a boolean "if aad_prefix field is set in the Thrift structure"..?

Reply:
Ok, I retract the question:) We need to differentiate between when is prefix is not used, and when used but not stored.